In [2]:
import pandas as pd 
import json

def load_data(file_name):
    data = []
    with open(file_name, 'r') as fin:
        for l in fin:
            d = json.loads(l)
            data.append(d)
    return data

In [4]:
df_children_books = load_data('goodreads_books_children.json')
df_children_books = pd.DataFrame(df_children_books)

df_reviews = load_data('goodreads_reviews_children.json')
df_reviews = pd.DataFrame(df_reviews)

df_interactions_filtered = pd.read_csv('df_interactions_filtered.csv')

# Data Preprocessing

In [5]:
df_children_books.dtypes

isbn                    object
text_reviews_count      object
series                  object
country_code            object
language_code           object
popular_shelves         object
asin                    object
is_ebook                object
average_rating          object
kindle_asin             object
similar_books           object
description             object
format                  object
link                    object
authors                 object
publisher               object
num_pages               object
publication_day         object
isbn13                  object
publication_month       object
edition_information     object
publication_year        object
url                     object
image_url               object
book_id                 object
ratings_count           object
work_id                 object
title                   object
title_without_series    object
dtype: object

In [6]:
df_interactions_filtered.dtypes

user_id                   object
book_id                    int64
review_id                 object
is_read                     bool
rating                     int64
review_text_incomplete    object
date_added                object
date_updated              object
read_at                   object
started_at                object
dtype: object

In [7]:
df_reviews.dtypes

user_id         object
book_id         object
review_id       object
rating           int64
review_text     object
date_added      object
date_updated    object
read_at         object
started_at      object
n_votes          int64
n_comments       int64
dtype: object

In [8]:
# Converting columns to the correct data type
integer_columns = ['text_reviews_count', 'num_pages', 'publication_day', 'publication_month', 'publication_year', 'ratings_count', 'rating', 'n_votes', 'n_comments', 'average_rating']
boolean_columns = ['is_ebook', 'is_read']
list_columns = ['series', 'popular_shelves', 'similar_books', 'authors']
datetime_columns = ['date_added', 'date_updated', 'read_at', 'started_at']

In [9]:
def convert_datatype(df):
    # Convert integer columns that exist in the DataFrame
    df_integer_columns = [col for col in integer_columns if col in df.columns]
    df[df_integer_columns] = df[df_integer_columns].apply(pd.to_numeric, errors='coerce', downcast='integer')
    
    # Convert boolean columns that exist in the DataFrame
    df_boolean_columns = [col for col in boolean_columns if col in df.columns]
    df[df_boolean_columns] = df[df_boolean_columns].map(lambda x: x.lower() == 'true')
    
    # Convert datetime columns that exist in the DataFrame
    df_datetime_columns = [col for col in datetime_columns if col in df.columns]
    df[df_datetime_columns] = df[df_datetime_columns].apply(pd.to_datetime, errors='coerce')

    # Convert list col
    df_list_columns = [col for col in list_columns if col in df.columns]
    
    # Convert remaining columns to string type
    remaining_columns = df.columns.difference(df_integer_columns + df_boolean_columns + df_datetime_columns + df_list_columns)
    df[remaining_columns] = df[remaining_columns].astype(str)

    return df


In [10]:
df_children_books = convert_datatype(df_children_books)

In [11]:
df_children_books.dtypes

isbn                     object
text_reviews_count        int32
series                   object
country_code             object
language_code            object
popular_shelves          object
asin                     object
is_ebook                   bool
average_rating          float64
kindle_asin              object
similar_books            object
description              object
format                   object
link                     object
authors                  object
publisher                object
num_pages               float64
publication_day         float64
isbn13                   object
publication_month       float64
edition_information      object
publication_year        float64
url                      object
image_url                object
book_id                  object
ratings_count             int32
work_id                  object
title                    object
title_without_series     object
dtype: object

In [12]:
# Convert book_id back to string in df_interactions_filtered
df_interactions_filtered['book_id'] = df_interactions_filtered['book_id'].astype(str)

In [13]:
# Convert datetime columns in df_interactions_filtered
df_interactions_filtered[datetime_columns] = df_interactions_filtered[datetime_columns].apply(
    pd.to_datetime, format='%a %b %d %H:%M:%S %z %Y', errors='coerce'
)

C:\Users\rayna\AppData\Local\Temp\ipykernel_23084\810077808.py:2: FutureWarning: In a future version of pandas, parsing datetimes with mixed time zones will raise an error unless `utc=True`. Please specify `utc=True` to opt in to the new behaviour and silence this warning. To create a `Series` with mixed offsets and `object` dtype, please use `apply` and `datetime.datetime.strptime`
  df_interactions_filtered[datetime_columns] = df_interactions_filtered[datetime_columns].apply(
C:\Users\rayna\AppData\Local\Temp\ipykernel_23084\810077808.py:2: FutureWarning: In a future version of pandas, parsing datetimes with mixed time zones will raise an error unless `utc=True`. Please specify `utc=True` to opt in to the new behaviour and silence this warning. To create a `Series` with mixed offsets and `object` dtype, please use `apply` and `datetime.datetime.strptime`
  df_interactions_filtered[datetime_columns] = df_interactions_filtered[datetime_columns].apply(
C:\Users\rayna\AppData\Local\Temp\

In [14]:
df_interactions_filtered.dtypes

user_id                   object
book_id                   object
review_id                 object
is_read                     bool
rating                     int64
review_text_incomplete    object
date_added                object
date_updated              object
read_at                   object
started_at                object
dtype: object

In [15]:
# Convert datetime columns in df_reviews
df_reviews[datetime_columns] = df_reviews[datetime_columns].apply(
    pd.to_datetime, format='%a %b %d %H:%M:%S %z %Y', errors='coerce'
)

C:\Users\rayna\AppData\Local\Temp\ipykernel_23084\3094010240.py:2: FutureWarning: In a future version of pandas, parsing datetimes with mixed time zones will raise an error unless `utc=True`. Please specify `utc=True` to opt in to the new behaviour and silence this warning. To create a `Series` with mixed offsets and `object` dtype, please use `apply` and `datetime.datetime.strptime`
  df_reviews[datetime_columns] = df_reviews[datetime_columns].apply(
C:\Users\rayna\AppData\Local\Temp\ipykernel_23084\3094010240.py:2: FutureWarning: In a future version of pandas, parsing datetimes with mixed time zones will raise an error unless `utc=True`. Please specify `utc=True` to opt in to the new behaviour and silence this warning. To create a `Series` with mixed offsets and `object` dtype, please use `apply` and `datetime.datetime.strptime`
  df_reviews[datetime_columns] = df_reviews[datetime_columns].apply(
C:\Users\rayna\AppData\Local\Temp\ipykernel_23084\3094010240.py:2: FutureWarning: In a f

In [16]:
df_reviews.dtypes

user_id         object
book_id         object
review_id       object
rating           int64
review_text     object
date_added      object
date_updated    object
read_at         object
started_at      object
n_votes          int64
n_comments       int64
dtype: object

# EDA

In [44]:
# Function to summarize data in dataset
def summarize_df(df):
    print(f'Shape: {df.shape[0]} rows {df.shape[1]} columns')
    for col in df.columns:
        print(f'Column - {col}, Data type - {df[col].dtype}')
        # Numerical
        if col in integer_columns:
            print('Number of NA Values:', sum(df[col].isnull()))
            print('Min value:', min(df[col]), ', Max value:', max(df[col]), ', Average value:', df[col].mean())

        # Boolean
        elif col in boolean_columns:
            print('Number of NA Values:', sum(df[col].isnull()))
            print(df.groupby(col).size())

        # List
        elif col in list_columns:
            print('Number of empty lists or NA Values:', df[col].isnull().sum() + sum(len(item) == 0 for item in df[col] if isinstance(item, list)))
            # print('Number of empty lists:', sum(len(item) == 0 for item in df[col]))

        # Datetime
        elif col in datetime_columns:
            print('Number of NA Values:', sum(df[col].isnull()))
            print('Min value:', min(df[col]), ', Max value:', max(df[col]))

        # String
        else:
            print('Number of NA Values:', sum((df[col].str.strip() == '') | (df[col].isnull())))
            if len(df[col].unique()) < 5:
                print('Unique Values:', df[col].unique())
        
        print("-"*40)

In [45]:
summarize_df(df_children_books)

Shape: 124082 rows 29 columns
Column - isbn, Data type - object
Number of NA Values: 20199
----------------------------------------
Column - text_reviews_count, Data type - int32
Number of NA Values: 0
Min value: 0 , Max value: 49850 , Average value: 27.08620911977563
----------------------------------------
Column - series, Data type - object
Number of empty lists or NA Values: 81213
----------------------------------------
Column - country_code, Data type - object
Number of NA Values: 0
Unique Values: ['US']
----------------------------------------
Column - language_code, Data type - object
Number of NA Values: 72496
----------------------------------------
Column - popular_shelves, Data type - object
Number of empty lists or NA Values: 0
----------------------------------------
Column - asin, Data type - object
Number of NA Values: 117363
----------------------------------------
Column - is_ebook, Data type - bool
Number of NA Values: 0
is_ebook
False    114606
True       9476
dtype

### Column Removal Justification for df_children_books
In our analysis of the df_children_books DataFrame, which contains 124,082 rows and 29 columns, we identified specific columns to remove based on their relevance and usefulness for building a recommendation system for books. Below are the columns we decided to exclude, along with the reasons for their removal:

- Insufficient Data: The `series` column had a high number of empty lists and/or null values (81,213 out of 124,082 rows), indicating insufficient data for meaningful analysis.
- Redundancy: Columns like `asin`, `kindle_asin`, and `work_id` were deemed redundant since the isbn column could serve as a unique identifier for each book.
- Uninformative: Columns such as `country_code`, `link`, `url`, and `image_url` do not provide useful information for our recommendation model
- Prioritization: Between `publisher` and `authors`, we retained the `authors` column as we believed that authors are more likely to influence the similarity between books than publishers and between title and title_without_series we retained the `title` column instead of the `title_without_series` column, as the full title is more informative.
- Time granularity: For publication dates, we retained the `publication_year` column as we deemed the day and month less significant for recommendations.

In [58]:
df_children_books_final = df_children_books[['isbn', 'isbn13', 'book_id', 'title', 'authors', 'num_pages', 'language_code', 'description', 'format', 'is_ebook', 'popular_shelves', 'average_rating', 'ratings_count', 'text_reviews_count']]
print(df_children_books_final.shape)
df_children_books_final.head(5)

(124082, 14)


isbn         isbn13  book_id  \
0  1599150603  9781599150604   287141   
1  1934876569  9781934876565  6066812   
2  0590417010  9780590417013    89378   
3  0915190575  9780915190577  3209312   
4  1416904999  9781416904991  1698376   

                                               title  \
0                      The Aeneid for Boys and Girls   
1  All's Fairy in Love and War (Avalon: Web of Ma...   
2                                         Dog Heaven   
3  Moths and Mothers, Feathers and Fathers: A Sto...   
4                                    What Do You Do?   

                                             authors num_pages language_code  \
0             [{'author_id': '3041852', 'role': ''}]       162                 
1               [{'author_id': '19158', 'role': ''}]       216                 
2                [{'author_id': '5411', 'role': ''}]        40           eng   
3  [{'author_id': '589328', 'role': ''}, {'author...                           
4              [{'author_id': '169159', 'role': ''}]        24                 

                                         description      format is_ebook  \
0  Relates in vigorous prose the tale of Aeneas, ...   Paperback    false   
1  To Kara's astonishment, she discovers that a p...   Paperback    false   
2  In Newbery Medalist Cynthia Rylant's classic b...   Hardcover    false   
3                                                                   false   
4  WHAT DO YOU DO?\nA hen lays eggs...\nA cow giv...  Board Book    false   

                                     popular_shelves average_rating  \
0  [{'count': '56', 'name': 'to-read'}, {'count':...           4.13   
1  [{'count': '515', 'name': 'to-read'}, {'count'...           4.22   
2  [{'count': '450', 'name': 'to-read'}, {'count'...           4.43   
3  [{'count': '8', 'name': 'to-read'}, {'count': ...           4.29   
4  [{'count': '8', 'name': 'to-read'}, {'count': ...           3.57   

  ratings_count text_reviews_count  
0            46                  7  
1            98                  6  
2          1331                193  
3            11                  4  
4            23                  4

In [48]:
summarize_df(df_interactions_filtered)

Shape: 6626989 rows 10 columns
Column - user_id, Data type - object
Number of NA Values: 0
----------------------------------------
Column - book_id, Data type - object
Number of NA Values: 0
----------------------------------------
Column - review_id, Data type - object
Number of NA Values: 0
----------------------------------------
Column - is_read, Data type - bool
Number of NA Values: 0
is_read
True    6626989
dtype: int64
----------------------------------------
Column - rating, Data type - int64
Number of NA Values: 0
Min value: 0 , Max value: 5 , Average value: 3.873383372146838
----------------------------------------
Column - review_text_incomplete, Data type - object
Number of NA Values: 5908387
----------------------------------------
Column - date_added, Data type - object
Number of NA Values: 0
Min value: 1990-01-01 00:00:00-08:00 , Max value: 2017-11-03 23:37:07-07:00
----------------------------------------
Column - date_updated, Data type - object
Number of NA Values: 0

### Column Removal Justification for df_interactions_filtered
In our analysis of the df_interactions_filtered DataFrame, which contains 6,626,989 rows and 10 columns, we identified specific columns to remove based on their relevance and usefulness for building a recommendation system for books. Below are the columns we decided to exclude, along with the reasons for their removal:

- Insufficient Data: The `read_at` and `started_at` columns had a high number of null values (4,742,229 and 5,639,401 out of 6,626,989 rows respectively), indicating insufficient data for meaningful analysis.
- Redundancy: Retained the `date_added` column for possible recommendations based on recency and dropped the `date_updated` column.
- Uninformative: The `is_read` column was dropped as it only contains true values as we have previously filtered out the rows where is_read = False in the data.ipynb file.

In [60]:
df_interactions_final = df_interactions_filtered.drop(['is_read', 'date_updated', 'read_at', 'started_at'], axis=1)
df_interactions_final

user_id   book_id  \
0        8842281e1d1347389f2ab93d60773d4d  23310161   
1        8842281e1d1347389f2ab93d60773d4d  18296097   
2        8842281e1d1347389f2ab93d60773d4d    817720   
3        8842281e1d1347389f2ab93d60773d4d    502362   
4        8842281e1d1347389f2ab93d60773d4d   1969280   
...                                   ...       ...   
6626984  6a78dcbe905b4caf5566301989327547       378   
6626985  6a78dcbe905b4caf5566301989327547      3636   
6626986  037adfe7bd500ce1e85f8647e3bb26b6    233818   
6626987  037adfe7bd500ce1e85f8647e3bb26b6      3636   
6626988  d0f6d1a4edcab80a6010cfcfeda4999f      8127   

                                review_id  rating  \
0        f4b4b050f4be00e9283c92a814af2670       4   
1        bc9cff98f54be3b2b8c5b34598a7546c       5   
2        75fd46041466ceb406b7fd69b089b9c5       5   
3        be1ad51fa3d519e39050d2a61ffab534       5   
4        5809d5592ee32745e048a9c67ac27100       5   
...                                   ...     ...   
6626984  ed74a6ebfc19334b9d06878a97757e4a       4   
6626985  48a2e637d0df4edae03e6615fb4f1908       4   
6626986  2e031f4f9fb6c58ffb6847f2e34415d4       4   
6626987  4b516a0ac05ef8c3e489d2ce8c442b96       5   
6626988  d68a98975f63ef3f3015a7daadafbdcb       4   

              review_text_incomplete                      date_added  
0        Fun sequel to the original.  Tue Nov 17 11:37:35 -0800 2015  
1                                NaN  Mon Sep 21 08:16:57 -0700 2015  
2                                NaN  Wed May 20 21:29:23 -0700 2015  
3                                NaN  Tue Mar 31 20:00:12 -0700 2015  
4                                NaN  Sat Nov 08 08:56:58 -0800 2014  
...                              ...                             ...  
6626984                          NaN  Mon Jun 04 17:06:57 -0700 2012  
6626985                          NaN  Mon Jun 04 17:05:54 -0700 2012  
6626986                          NaN  Tue Feb 12 16:26:25 -0800 2013  
6626987                          NaN  Tue Feb 12 16:20:27 -0800 2013  
6626988                          NaN  Mon Jun 04 18:17:20 -0700 2012  

[6626989 rows x 6 columns]

In [47]:
summarize_df(df_reviews)

Shape: 734640 rows 11 columns
Column - user_id, Data type - object
Number of NA Values: 0
----------------------------------------
Column - book_id, Data type - object
Number of NA Values: 0
----------------------------------------
Column - review_id, Data type - object
Number of NA Values: 0
----------------------------------------
Column - rating, Data type - int64
Number of NA Values: 0
Min value: 0 , Max value: 5 , Average value: 3.818343678536426
----------------------------------------
Column - review_text, Data type - object
Number of NA Values: 158
----------------------------------------
Column - date_added, Data type - object
Number of NA Values: 0
Min value: 2000-08-15 00:00:00-07:00 , Max value: 2017-11-03 15:02:43-07:00
----------------------------------------
Column - date_updated, Data type - object
Number of NA Values: 0
Min value: 2006-08-29 10:18:37-07:00 , Max value: 2017-11-05 13:05:10-08:00
----------------------------------------
Column - read_at, Data type - obje

In [61]:
print(df_reviews.columns)

Index(['user_id', 'book_id', 'review_id', 'rating', 'review_text',
       'date_added', 'date_updated', 'read_at', 'started_at', 'n_votes',
       'n_comments'],
      dtype='object')


### Column Removal Justification for df_reviews
The df_reviews DataFrame contains 734,640 rows and 11 columns. We only retained the `review_id` and `n_votes` columns as we will be doing a left join on df_interactions_filtered to integrate the data from df_reviews to df_interactions_filtered. df_interactions_filtered contains interactions without text reviews, which could be valuable for recommendations. Thus, we decided to retain reviews without text as well. The `review_id` column can be used to uniquely identify reviews for the left join, while `n_votes` provides insight into the popularity or quality of a review.

In [62]:
df_reviews_final = df_reviews[['review_id', 'n_votes']]
df_reviews_final

review_id  n_votes
0       f4b4b050f4be00e9283c92a814af2670        7
1       22d424a2b0057b18fb6ecf017af7be92        4
2       50ed4431c451d5677d98dd25ca8ec106        6
3       1e4de11dd4fa4b7ffa59b6c69a6b28e9        5
4       2065145714bf747083a1c9ce81d5c4fe        5
...                                  ...      ...
734635  6251b6332ee9f898fe0e975821eb747e        0
734636  a4f8b16d5c0c5dc08d0f3fa118a262b4        0
734637  a5964e464961cf0185d31e00a7a9ffac        0
734638  42729a1e85ca16fa5e4a44ffccd020e4        0
734639  47658630f9a764af427068b618ecbf02        0

[734640 rows x 2 columns]

In [63]:
df_interactions_final_merged = df_interactions_final.merge(df_reviews_final, how='left')
df_interactions_final_merged

user_id   book_id  \
0        8842281e1d1347389f2ab93d60773d4d  23310161   
1        8842281e1d1347389f2ab93d60773d4d  18296097   
2        8842281e1d1347389f2ab93d60773d4d    817720   
3        8842281e1d1347389f2ab93d60773d4d    502362   
4        8842281e1d1347389f2ab93d60773d4d   1969280   
...                                   ...       ...   
6626984  6a78dcbe905b4caf5566301989327547       378   
6626985  6a78dcbe905b4caf5566301989327547      3636   
6626986  037adfe7bd500ce1e85f8647e3bb26b6    233818   
6626987  037adfe7bd500ce1e85f8647e3bb26b6      3636   
6626988  d0f6d1a4edcab80a6010cfcfeda4999f      8127   

                                review_id  rating  \
0        f4b4b050f4be00e9283c92a814af2670       4   
1        bc9cff98f54be3b2b8c5b34598a7546c       5   
2        75fd46041466ceb406b7fd69b089b9c5       5   
3        be1ad51fa3d519e39050d2a61ffab534       5   
4        5809d5592ee32745e048a9c67ac27100       5   
...                                   ...     ...   
6626984  ed74a6ebfc19334b9d06878a97757e4a       4   
6626985  48a2e637d0df4edae03e6615fb4f1908       4   
6626986  2e031f4f9fb6c58ffb6847f2e34415d4       4   
6626987  4b516a0ac05ef8c3e489d2ce8c442b96       5   
6626988  d68a98975f63ef3f3015a7daadafbdcb       4   

              review_text_incomplete                      date_added  n_votes  
0        Fun sequel to the original.  Tue Nov 17 11:37:35 -0800 2015      7.0  
1                                NaN  Mon Sep 21 08:16:57 -0700 2015      NaN  
2                                NaN  Wed May 20 21:29:23 -0700 2015      NaN  
3                                NaN  Tue Mar 31 20:00:12 -0700 2015      NaN  
4                                NaN  Sat Nov 08 08:56:58 -0800 2014      NaN  
...                              ...                             ...      ...  
6626984                          NaN  Mon Jun 04 17:06:57 -0700 2012      NaN  
6626985                          NaN  Mon Jun 04 17:05:54 -0700 2012      NaN  
6626986                          NaN  Tue Feb 12 16:26:25 -0800 2013      NaN  
6626987                          NaN  Tue Feb 12 16:20:27 -0800 2013      NaN  
6626988                          NaN  Mon Jun 04 18:17:20 -0700 2012      NaN  

[6626989 rows x 7 columns]

# Data Cleaning

## isbn and isbn13  
1. remove rows with both empty isbn and isbn13 
2. create new column called 'isbn_combined'. if isbn present, take value from isbn column, else, take value from isbn13 column. 

rationale: prep data for web scraping

In [64]:
print(sum((df_children_books_final['isbn13'] == '') & (df_children_books_final['isbn'] == '')))
print(sum((df_children_books_final['isbn'] == '')))
print(sum((df_children_books_final['isbn13'] == '')))

12924
20199
15489


In [65]:
df_children_books_final.drop(
    df_children_books_final[(df_children_books_final['isbn13'] == '') & (df_children_books_final['isbn'] == '')].index,
    inplace=True
)
df_children_books_final.shape

/var/folders/pj/nr405prs69j7fp08s_fm2qc40000gn/T/ipykernel_7275/2325941102.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_children_books_final.drop(


(111158, 14)

In [67]:

df_children_books_final['isbn_combined'] = df_children_books_final.apply(
    lambda row: row['isbn13'] if row['isbn'] == '' else row['isbn'], axis=1
)
print(sum(df_children_books_final['isbn_combined'] == ''))
df_children_books_final

0


/var/folders/pj/nr405prs69j7fp08s_fm2qc40000gn/T/ipykernel_7275/3581842822.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_children_books_final['isbn_combined'] = df_children_books_final.apply(


isbn         isbn13   book_id  \
0       1599150603  9781599150604    287141   
1       1934876569  9781934876565   6066812   
2       0590417010  9780590417013     89378   
3       0915190575  9780915190577   3209312   
4       1416904999  9781416904991   1698376   
...            ...            ...       ...   
124075              9781406368475  30164426   
124076  1575054035  9781575054032    823094   
124079  0061960314  9780061960314   7925060   
124080  0689852959  9780689852954    331839   
124081  0765197456  9780765197450   2342551   

                                                    title  \
0                           The Aeneid for Boys and Girls   
1       All's Fairy in Love and War (Avalon: Web of Ma...   
2                                              Dog Heaven   
3       Moths and Mothers, Feathers and Fathers: A Sto...   
4                                         What Do You Do?   
...                                                   ...   
124075      Flora and Ulysses: The Illuminated Adventures   
124076                     To Root, to Toot, to Parachute   
124079                                       Instructions   
124080     Jacqueline Kennedy Onassis: Friend of the Arts   
124081           The Children's Classic Poetry Collection   

                                                  authors num_pages  \
0                  [{'author_id': '3041852', 'role': ''}]       162   
1                    [{'author_id': '19158', 'role': ''}]       216   
2                     [{'author_id': '5411', 'role': ''}]        40   
3       [{'author_id': '589328', 'role': ''}, {'author...             
4                   [{'author_id': '169159', 'role': ''}]        24   
...                                                   ...       ...   
124075               [{'author_id': '13663', 'role': ''}]       251   
124076  [{'author_id': '44955', 'role': ''}, {'author_...             
124079  [{'author_id': '1221698', 'role': ''}, {'autho...        40   
124080  [{'author_id': '10681', 'role': ''}, {'author_...       176   
124081         [{'author_id': '82312', 'role': 'Editor'}]        96   

       language_code                                        description  \
0                     Relates in vigorous prose the tale of Aeneas, ...   
1                     To Kara's astonishment, she discovers that a p...   
2                eng  In Newbery Medalist Cynthia Rylant's classic b...   
3                                                                         
4                     WHAT DO YOU DO?\nA hen lays eggs...\nA cow giv...   
...              ...                                                ...   
124075           eng  Holy unanticipated occurrences! A cynic meets ...   
124076                Rhyming text and illustrations of comical cats...   
124079                "A perfect reminder to always be on the lookou...   
124080                One of the most popular series ever published ...   
124081                Gathers poems by William Blake, Emily Bronte, ...   

            format is_ebook  \
0        Paperback    false   
1        Paperback    false   
2        Hardcover    false   
3                     false   
4       Board Book    false   
...            ...      ...   
124075   Paperback    false   
124076   Hardcover    false   
124079   Hardcover    false   
124080   Paperback    false   
124081   Hardcover    false   

                                          popular_shelves average_rating  \
0       [{'count': '56', 'name': 'to-read'}, {'count':...           4.13   
1       [{'count': '515', 'name': 'to-read'}, {'count'...           4.22   
2       [{'count': '450', 'name': 'to-read'}, {'count'...           4.43   
3       [{'count': '8', 'name': 'to-read'}, {'count': ...           4.29   
4       [{'count': '8', 'name': 'to-read'}, {'count': ...           3.57   
...                                                   ...            ...   
124075  [{'count': '7743', 'name': 'to-read'}, {'count...

# Feature Engineering 

After clean data and web scrape for genre, do one hot encoding for genre. (another feature)

In [ ]:
def categorize_format(format_value):
    if 'ebook' in format_value.lower() or 'kindle' in format_value.lower() or 'pdf' in format_value.lower():
        return 'Digital'
    elif 'audio' in format_value.lower() or 'audiobook' in format_value.lower() or 'cd' in format_value.lower():
        return 'Audio'
    elif 'paper' in format_value.lower() or 'hardcover' in format_value.lower() or 'softcover' in format_value.lower():
        return 'Physical'
    elif 'board' in format_value.lower() or 'novelty' in format_value.lower():
        return 'Specialty'
    else:
        return 'Other'